Fix issue #120

mateuszbaran opened the pull request:
This change fixes issue #120.

grzanka commented:
It seems that the problem still persists, but manifest itself on smaller scale, and for lower energies:

![image](https://user-images.githubusercontent.com/7374733/106279606-f8ffb700-623c-11eb-922a-301fc4e73c61.png)

grzanka commented:
@mateuszbaran I suggest to check how accurate is the series expansion used in `AT_RDD_Katz_PowerLawER_DaverageKernel_approx`
https://github.com/libamtrack/library/blob/master/include/AT_RDD_ShellAveraged.h#L123
when compared to special function implemented in 
`AT_RDD_Katz_PowerLawER_DaverageKernel`
https://github.com/libamtrack/library/blob/master/include/AT_RDD_ShellAveraged.h#L91

Comparison should be made for low (i.e. 1MeV protons) and high (i.e. 150 MeV protons) energies

mateuszbaran commented:
For low energies the approximation in `AT_RDD_Katz_PowerLawER_DaverageKernel_approx` isn't accurate enough, I'll write a more accurate approximation.

grzanka commented:
Another idea - let us use another special function, i.e. beta:

https://www.wolframalpha.com/input/?i=int+1%2Fx+*+1%2Falpha+*+%281+-+x%29%5E%7B1%2Falpha%7D+dx
![image](https://user-images.githubusercontent.com/7374733/106327181-f70a1800-627d-11eb-9575-8808d2a16b2a.png)


It seems to have some reasonable values, even for arguments close to 1:
https://www.wolframalpha.com/input/?i=beta%281+-+1e-8%2C+1+%2B+1%2F1.666%2C+0%29
![image](https://user-images.githubusercontent.com/7374733/106327270-12752300-627e-11eb-879b-8d49d11cf2f2.png)

Following description from here: https://en.wikipedia.org/wiki/Beta_function#Properties
regularized beta function could be implemented easily.

Unfortunately gsl implementation (gsl_sf_beta_inc) works only for b > 0.

mateuszbaran commented:
I've managed to get some nice approximations from Wolfram Alpha and I have obtained a simple code for this function:
```julia
using SpecialFunctions

"""
    be(x, a)

Beta_x(a, 0)
"""
function be(x, a)
    if x < 0.9
        N = 10^3
        ret = 0.0
        for i in 0:N
            ret += x^i/(a+i)
        end
        ret *= x^a
        return ret
    else
        -digamma(a) - log(1-x) - Base.MathConstants.eulergamma + a*(1-x)*(digamma(a) - digamma(a+1) + 1)
    end
end
```

I need to tweak it still but it gives very similar results to wolfram alpha for some sample points.

grzanka commented:
> I've managed to get some nice approximations from Wolfram Alpha and I have obtained a simple code for this function:
> 
> ```julia
> using SpecialFunctions
> 
> """
>     be(x, a)
> 
> Beta_x(a, 0)
> """
> function be(x, a)
>     if x < 0.9
>         N = 10^3
>         ret = 0.0
>         for i in 0:N
>             ret += x^i/(a+i)
>         end
>         ret *= x^a
>         return ret
>     else
>         -digamma(a) - log(1-x) - Base.MathConstants.eulergamma + a*(1-x)*(digamma(a) - digamma(a+1) + 1)
>     end
> end
> ```
> 
> I need to tweak it still but it gives very similar results to wolfram alpha for some sample points.

Looks nice.  Can you change the way you calculate series expansion (`x^i/(a+i)`) ? Instead of large number of iterations there should be some condition for convergence. 

I see that the implementation of digamma function in GSL is based on some interpolation polynomials - can you take a look on that and see what is the accuracy for our parameters (i.e. alpha) ?
https://github.com/ampl/gsl/blob/dee1063dd57ec2841981b010eb732901b452c30a/specfunc/psi.c

Finally - can you provide in the issue or PR comments some piece of the code which does a cross-check with values calculated using other methods ?
Python mpmath library has some working implementation:
https://mpmath.org/doc/current/functions/gamma.html#betainc
![image](https://user-images.githubusercontent.com/7374733/106351917-90b5e180-62df-11eb-9261-9bedf51ba8ea.png)

Python code here:
https://github.com/fredrik-johansson/mpmath/blob/77c4c5e0ce37a2acca27bbf059e508bcb9579005/mpmath/functions/expintegrals.py#L109

Interesting - in most codes I've seen (mpmath in python and arb in C, both authored by fredrik-johansson) incomplete beta function for b=0 is implemented using 2F1 (hypergeometric function). Take a look here:
https://arblib.org/acb_hypgeom.html#incomplete-gamma-and-beta-functions (`acb_hypgeom_beta_lower`)

This brings us back to original idea - to use 2F1 instead of beta function.
@mateuszbaran do you have preference on which of them is better ?

Another issue: implementation of `AT_RDD_Katz_PowerLawER_DaverageKernel` seems to be completely broken. I suspect there should `1-x` instead of `1/x` as last argument.

mateuszbaran commented:
> Looks nice. Can you change the way you calculate series expansion (`x^i/(a+i)`) ? Instead of large number of iterations there should be some condition for convergence.

Yes, I'll change the loop termination condition to something more appropriate. It was just an initial sketch to see if this approach works.

> I see that the implementation of digamma function in GSL is based on some interpolation polynomials - can you take a look on that and see what is the accuracy for our parameters (i.e. alpha) ?
> https://github.com/ampl/gsl/blob/dee1063dd57ec2841981b010eb732901b452c30a/specfunc/psi.c

Yes, I can check that. What do you think would be the sufficient level of accuracy for the entire integral we calculate here? Probably something like <1e-10 relative error over x in [1e-12, 1] for alpha between 1 and 2?

> Finally - can you provide in the issue or PR comments some piece of the code which does a cross-check with values calculated using other methods ?
> Python mpmath library has some working implementation:
> https://mpmath.org/doc/current/functions/gamma.html#betainc

Sure, I will check the accuracy against mpmath.

> This brings us back to original idea - to use 2F1 instead of beta function.
> @mateuszbaran do you have preference on which of them is better ?

I think using either 2F1 or beta only make sense when we have access (in the library) to an implementation that handles our parameter range, and our integral is a very special case for both and isn't handles by GSL. I would prefer to just polish my approximation and check its accuracy against some other library like mpmath, then we can just use it (instead of adding a new dependency to libamtrack or trying to push this change to GSL itself).

grzanka commented:
> > I see that the implementation of digamma function in GSL is based on some interpolation polynomials - can you take a look on that and see what is the accuracy for our parameters (i.e. alpha) ?
> > https://github.com/ampl/gsl/blob/dee1063dd57ec2841981b010eb732901b452c30a/specfunc/psi.c
> 
> Yes, I can check that. What do you think would be the sufficient level of accuracy for the entire integral we calculate here? Probably something like <1e-10 relative error over x in [1e-12, 1] for alpha between 1 and 2?

Relative error < 1e-10 seems very reasonable estimate. 
For alpha we have just two values right now, so this should be quite straightforward.
The range of `x` should be more carefully analyzed, taking into account possible values of delta-electron range, lower cutoff for D(r) and radius at which we evaluate D(r).

There will be following applications for "shell averaging":
  - calculation of "plateau" of extended RDD, there we use x1=0 (but in fact we could use some larger values, as D(r) = 0 for r < r_min).
  - checking if D(r) (simple or point-like, not extended) integrated up to Rmax reproduces LET values. There we need x1=0 (or rmin) and x2 corresponding to rmax (range of delta electrons)
  - estimating how much dose is delivered within some limit (on radius) from ion track. There we need x1=0 (or rmin) and some intermediate values on x2 (probably corresponding to 0.1-0.5 of rmax).

> > This brings us back to original idea - to use 2F1 instead of beta function.
> > @mateuszbaran do you have preference on which of them is better ?
> 
> I think using either 2F1 or beta only make sense when we have access (in the library) to an implementation that handles our parameter range, and our integral is a very special case for both and isn't handles by GSL. I would prefer to just polish my approximation and check its accuracy against some other library like mpmath, then we can just use it (instead of adding a new dependency to libamtrack or trying to push this change to GSL itself).

I agree that using unstable implementation or adding another dependencies doesn't make much sense. I was just curious if there is any implementation on the market which we can borrow (i.e. translate to C). What I was considering was implementation of 2F1 in mpmath/arb packages. It seems to be stable (at least in python), but I haven't estimated effort needed to borrow it (or translate to C). Have you tried to understand how does it work under the hood ?

mateuszbaran commented:
> > 
> 
> I agree that using unstable implementation or adding another dependencies doesn't make much sense. I was just curious if there is any implementation on the market which we can borrow (i.e. translate to C). What I was considering was implementation of 2F1 in mpmath/arb packages. It seems to be stable (at least in python), but I haven't estimated effort needed to borrow it (or translate to C). Have you tried to understand how does it work under the hood ?

I did look at a couple packages that do this and 2F1 tends to be a maze of special cases. It turned out to be easier to do it from scratch than following all these transformations and branches.

mateuszbaran commented:
I've improved my previous version:
```C
double beta0(double x, double a) {
    if (x < 0.99) {
        double ret = 0.0;
        double xi = 1.0;
        long i = 0;
        xi = 1.0;
        while (xi > 1e-15) {
            ret += xi / (a + i);
            xi *= x;
            i += 1;
        }
        ret *= pow(x, a);
        return ret;
    }
    else {
        // https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
        const double psi_a = gsl_sf_psi(a);
        const double psi_a_1 = gsl_sf_psi(a + 1.0);
        const double psi_a_2 = gsl_sf_psi(a + 2.0);
        const double psi_a_3 = gsl_sf_psi(a + 3.0);
        return -psi_a - log(1.0-x) - M_EULER +
                a*(1.0-x)*(psi_a - psi_a_1 + 1.0) -
                0.25 * gsl_pow_2(1-x) * (
                        a * (a - 4*a*psi_a_1 + 2*a*psi_a_2 + 2 * (a-1)*psi_a + 2*psi_a_2 - 3)) +
                (1.0/18.0)*a*gsl_pow_3(1-x) * (
                        a*a + 9*a*a * psi_a_2 - 3*a*a*psi_a_3 + 3*(a*a - 3*a + 2)*psi_a - 6*a -
                        9*(a-1)*a*psi_a_1 + 9*a*psi_a_2 - 9*a*psi_a_3 - 6*psi_a_3 + 11);
    }
}

double integrate(double x1, double x2, double alpha) {
    return beta0(1-x1, 1+1/alpha) - beta0(1-x2, 1+1/alpha);
}

void evaluate_integration() {
    FILE* frefvals = fopen("reference_int_vals.txt", "r+");
    int N;
    double alpha;
    fscanf(frefvals, "%d", &N);
    fscanf(frefvals, "%lf", &alpha);
    double * xs = malloc(sizeof(double) * N);
    double * refs = malloc(sizeof(double) * N);
    for(long i=0; i<N; ++i) {
        fscanf(frefvals, "%lf", xs + i);
        fscanf(frefvals, "%lf", refs + i);
    }

    clock_t start, end;
    double max_error = 0.0;
    int max_error_at = 0;
    start = clock();
    for(long i=0; i<N; ++i) {
        double my_val = integrate(xs[i], 1.0, alpha);
        double ref_val = refs[i];
        double cur_error = fabs(my_val - ref_val)/ref_val;
        if(cur_error > max_error) {
            max_error = cur_error;
            max_error_at = i;
        }
    }
    end = clock();
    double cpu_time_used = 1e6 * ((double) (end - start)) / (CLOCKS_PER_SEC * N);
    printf("Average time taken per evaluation: %lf [us]\n", cpu_time_used);

    printf("Max error: %e at xs[%d] = %.10f\n",
           max_error,
           max_error_at,
           xs[max_error_at]);
    printf("We at max error: %.10e, ref at max error: %.10e\n",
           integrate(xs[max_error_at], 1.0, alpha),
           refs[max_error_at]);
    free(xs);
    free(refs);
}
```
I've compared it to results generated by the following script:
```Python
def calc_int(x, alpha):
    return betainc(a=1.0+1.0/alpha, b=0.0, x1=0, x2=1-x)


def save_calc_int(alpha):
    mp.dps = 25; mp.pretty = True
    N = 100000
    with open("reference_int_vals.txt", "w+") as f:
        f.write(str(N))
        f.write("\n")
        f.write(str(alpha))
        f.write("\n")
        for x in np.linspace(1e-12, 1.0, N):
            f.write(str(x))
            f.write("\t")
            f.write(str(calc_int(x, alpha)))
            f.write("\n")


save_calc_int(1.667)
```
and the maximum (relative) error is 2.152589e-11 (at x=0.0099900999). It looks like the part where the formula with `psi` gets not accurate enough is the hardest to calculate. I think this should be good enough for our needs.

grzanka commented:
> I've improved my previous version:
> 
> ```c
> double beta0(double x, double a) {
>     if (x < 0.99) {

Can you add some comment to the code, explaining the choice of the limit.
Imagine a reader coming back to this code few year from now.
He will be curious why series expansions wasn't used for all possible x values (I know its obvious for mathematician, but not good comments should also explain obvious things, as the reader may be a novice).

Next - why 0.99 and not 0.999 - this should also be explained.

>         double ret = 0.0;
>         double xi = 1.0;
>         long i = 0;
>         xi = 1.0;
>         while (xi > 1e-15) {

`1e-15` should be extracted  to a parameter of the funcion, in a similar way as integration methods in GSL do it:
https://www.gnu.org/software/gsl/doc/html/integration.html#examples

>             ret += xi / (a + i);
>             xi *= x;
>             i += 1;
>         }
>         ret *= pow(x, a);
>         return ret;
>     }
>     else {
>         // https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
>         const double psi_a = gsl_sf_psi(a);
>         const double psi_a_1 = gsl_sf_psi(a + 1.0);
>         const double psi_a_2 = gsl_sf_psi(a + 2.0);
>         const double psi_a_3 = gsl_sf_psi(a + 3.0);

How costly is evaluation of `gsl_sf_psi` ? 
In our use case `a` takes only one of two possible values. This function will be evaluated many times for different values of `x`, hence we will have multiple redundant calls to gsl_sf_psi. We should check it and consider later moving this part to precalculated data structure outside this function.

>         return -psi_a - log(1.0-x) - M_EULER +
>                 a*(1.0-x)*(psi_a - psi_a_1 + 1.0) -
>                 0.25 * gsl_pow_2(1-x) * (
>                         a * (a - 4*a*psi_a_1 + 2*a*psi_a_2 + 2 * (a-1)*psi_a + 2*psi_a_2 - 3)) +
>                 (1.0/18.0)*a*gsl_pow_3(1-x) * (
>                         a*a + 9*a*a * psi_a_2 - 3*a*a*psi_a_3 + 3*(a*a - 3*a + 2)*psi_a - 6*a -
>                         9*(a-1)*a*psi_a_1 + 9*a*psi_a_2 - 9*a*psi_a_3 - 6*psi_a_3 + 11);

Is it part of series expansion ? If yes, then it should be documented. Also write down some arguments why this many terms has been choosen.

> and the maximum (relative) error is 2.152589e-11 (at x=0.0099900999). It looks like the part where the formula with `psi` gets not accurate enough is the hardest to calculate. I think this should be good enough for our needs.

Yes, 1e-11 should be enough !

grzanka commented:
>         for x in np.linspace(1e-12, 1.0, N):

I think the check should be done for geometric series, not for arithmetic one. With your implementation you skip making checks at values of 1e-11 and 1e-10 if I am correct. 

>             f.write(str(x))
>             f.write("\t")
>             f.write(str(calc_int(x, alpha)))
>             f.write("\n")
> 
> 
> save_calc_int(1.667)
> ```
> 
> and the maximum (relative) error is 2.152589e-11 (at x=0.0099900999). It looks like the part where the formula with `psi` gets not accurate enough is the hardest to calculate. I think this should be good enough for our needs.

grzanka commented:
@mateuszbaran can you also make a proper commit with the proposed code ? It looks really reasonable

mateuszbaran commented:
Here is a new code. I'll put it in a commit soon. I didn't add documentation to `integrate` because it will just replace `AT_RDD_Katz_PowerLawER_DaverageKernel`.

```C
/**
 * Calculate the integral int_x^1 ((1-t)^(a-1))/t dt.
 * The accuacy was estimated at
 *
 * @param[in] x                     lower limit of integration
 * @param[in] a                     used in exponential of integrated function
 * @param[in] iteration_threshold   used for termination condition of a loop
 * @return          calculated value of the integral
 */
double beta0(double x, double a, double iteration_threshold) {
    if (x < 0.99) {
        // This iterative algorithms computes the value of the beta function very accurately
        // but converges very slowly when the value of x is very close to 1.
        double ret = 0.0;
        double xi = 1.0;
        long i = 0;
        while (xi > iteration_threshold) {
            ret += xi / (a + i);
            xi *= x;
            i += 1;
        }
        ret *= pow(x, a);
        return ret;
    }
    else {
        // Series expansion taken from https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
        // This can be obtained by Taylor expansion of the calculated integral + log(x) at x = 0.
        // Adding more terms didn't improve the accuracy.
        const double psi_a = gsl_sf_psi(a);
        const double psi_a_1 = psi_a + 1/a;
        const double psi_a_2 = psi_a_1 + 1/(a+1);
        const double psi_a_3 = psi_a_2 + 1/(a+2);
        const double a2 = a*a;
        return -psi_a - log(1.0-x) - M_EULER +
                a*(1.0-x)*(psi_a - psi_a_1 + 1.0) -
                0.25 * gsl_pow_2(1-x) * (
                        a * (a - 4*a*psi_a_1 + 2*a*psi_a_2 + 2 * (a-1)*psi_a + 2*psi_a_2 - 3)) +
                (1.0/18.0)*a*gsl_pow_3(1-x) * (
                        a2 + 9*a2 * psi_a_2 - 3*a2*psi_a_3 + 3*(a2 - 3*a + 2)*psi_a - 6*a -
                        9*(a-1)*a*psi_a_1 + 9*a*psi_a_2 - 9*a*psi_a_3 - 6*psi_a_3 + 11);
    }
}

double integrate(double x1, double x2, double alpha) {
    double iteration_threshold = 1e-13;
    return beta0(1-x1, 1+1/alpha, iteration_threshold) - beta0(1-x2, 1+1/alpha, iteration_threshold);
}

```

> Can you add some comment to the code, explaining the choice of the limit.
> Imagine a reader coming back to this code few year from now.
> He will be curious why series expansions wasn't used for all possible x values (I know its obvious for mathematician, but not good comments should also explain obvious things, as the reader may be a novice).
> 
> Next - why 0.99 and not 0.999 - this should also be explained.

It's basically a tradeoff: iterative formula is very accurate but increasingly slow at values of `x` close to 1. For example, evaluation using loop at x=0.99 takes about 10 microseconds, while evaluation using that expansion only about 88ns (and it could be even faster with some caching). How would you select this threshold?

> `1e-15` should be extracted to a parameter of the funcion, in a similar way as integration methods in GSL do it:
> https://www.gnu.org/software/gsl/doc/html/integration.html#examples

Should I make a separate struct for this parameter? For now I've just added it as a normal parameter of type `double`.

> How costly is evaluation of `gsl_sf_psi` ?
> In our use case `a` takes only one of two possible values. This function will be evaluated many times for different values of `x`, hence we will have multiple redundant calls to gsl_sf_psi. We should check it and consider later moving this part to precalculated data structure outside this function.

Entire evaluation of `beta0` with this approximation takes < 90 ns. I doubt improving performance of this part is particularly important?

> Is it part of series expansion ? If yes, then it should be documented. Also write down some arguments why this many terms has been choosen.

I've added some documentation for it.

> I think the check should be done for geometric series, not for arithmetic one. With your implementation you skip making checks at values of 1e-11 and 1e-10 if I am correct.

I've checked on both, and anyway the accuracy is worst for the largest values of `x` handled by that `else` branch.

> @mateuszbaran can you also make a proper commit with the proposed code ? It looks really reasonable

Sure, I'll make one soon.

mateuszbaran commented:
I've pushed a new commit with this more accurate approximation. It looks like there is no noticeable jump near 1e-9 now.

grzanka commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
+    else {
+        // Series expansion taken from https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
+        // This can be obtained by Taylor expansion of the calculated integral + log(x) at x = 0.
+        // Adding more terms didn't improve the accuracy.
```
Add some explanations why we do not use 2F1 function as wolfram alpha suggests.

mateuszbaran replied:
I've added an explanation to the header of the `beta0` function.

grzanka commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
+    }
+    else {
+        // Series expansion taken from https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
+        // This can be obtained by Taylor expansion of the calculated integral + log(x) at x = 0.
```
I do not get why `+ log(x)` is used.

mateuszbaran replied:
I've added an explanation in the new commit. Alternatively, I could explain it in terms of adding 1/x to the integrand, I'm not sure which one would be more clear.

grzanka commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
+        return ret;
+    }
+    else {
+        // Series expansion taken from https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx
```
Wolfram alpha provides series expansion at x = 0. Here we need series expansion for x = 1.
Explain in comment how did you get from one case to another one.

mateuszbaran replied:
I've added an explanation in the new commit.

grzanka commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
-  const double C2 = pow( 1.0-x2, alpha_inv ) * pow( (x2-1.0)/x2, -alpha_inv );
-  const double HGF2 = gsl_sf_hyperg_2F1( -alpha_inv, -alpha_inv, 1.0-alpha_inv, 1.0/x2 );
-  return 2.0 * (C2 * HGF2 - C1 * HGF1) / (gsl_pow_2(x2) - gsl_pow_2(x1));
+  // this threshold was selected to get the overall error < 1e-10;
```
Please update comments in header file for method `AT_RDD_Katz_PowerLawER_DaverageKernel`

mateuszbaran replied:
I've updated it in the new commit.

mateuszbaran commented on include/AT_RDD_ShellAveraged.h:
```diff
@@ -56,6 +56,17 @@
+ * @param[in] iteration_threshold   used for termination condition of a loop
+ * @return          calculated value of the integral
+ */
+double beta0(double x, double a, double iteration_threshold);
```
Wolfram alpha considers it to be special case of the incomplete beta function: https://www.wolframalpha.com/input/?i=int+-1%2Fx+%281+-+x%29%5E%28a-1%29+dx but I don't know how widespread that definition is. Usually the incomplete beta function is only defined for a, b > 0. mpmath defines the incomplete beta function a bit differently (with both limits of integration as parameters), so I could use it to generate reference values.

mateuszbaran commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
 
+double   beta0(double x, double a, double iteration_threshold) {
+    if (x < 0.99) {
+        // This iterative algorithms computes the value of the beta function very accurately
```
I've added an explanation about this algorithm.

mateuszbaran commented on include/AT_RDD_ShellAveraged.h:
```diff
@@ -56,6 +56,17 @@
 
+/**
+ * Calculate the integral int_x^1 ((1-t)^(a-1))/t dt.
+ * The accuacy was estimated at
```
I've filled the missing part.

mateuszbaran commented on src/AT_RDD_ShellAveraged.c:
```diff
@@ -41,22 +43,49 @@
   return 2.0 * Katz_point_coeff_Gy * ( log(r2_m/r1_m) - (r2_m - r1_m)/max_electron_range_m ) / (gsl_pow_2(r2_m/max_electron_range_m) - gsl_pow_2(r1_m/max_electron_range_m));
 }
 
+double   beta0(double x, double a, double iteration_threshold) {
```
I've done it in the new commit.

mateuszbaran commented on include/AT_RDD_ShellAveraged.h:
```diff
@@ -56,6 +56,17 @@
+ * @param[in] iteration_threshold   used for termination condition of a loop
+ * @return          calculated value of the integral
+ */
+double beta0(double x, double a, double iteration_threshold);
```
I've done it in the new commit.

mateuszbaran commented on include/AT_RDD_ShellAveraged.h:
```diff
@@ -56,6 +56,17 @@
 /* --------------------------------------------------- SHELL AVERAGE DOSE ---------------------------------------------------*/
 
+/**
+ * Calculate the integral int_x^1 ((1-t)^(a-1))/t dt.
```
I've done it in the new commit.
